flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -876,6 +887,17 @@ private void jobStatusChanged(
 		validateRunsInMainThread();
 
 		if (newJobStatus.isGloballyTerminalState()) {
+			// other terminal job states are handled by the executions
+			if (newJobStatus == JobStatus.FINISHED) {
```

Comment:
From my understanding, I think we can only consider FINISHED due to that when transit to CANCELED and FAILED , all execution vertices will be canceled and the thus the result partition will be canceled via `sendReleaseIntermediateResultPartitionsRpcCall`. 

However, with region failover, an execution may need to re-run after it turns into FINISHED. This happens for cases like `A -> C, B->C`, A and B are all finished but C fails to read data from A, then B also need to re-execute if the partition is random (like RebalancePartition). When the execution vertex is `resetForNewExecution`, the result partition of previous execution is not released. Then when we `cancel` or `failGlobal` on the ExecutionGraph, it will also only consider the last execution, then some result partitions may be leaked.

Reply:
failover scenarios are not covered by this PR; it only adds a bare-bones implementation for cleaning up after the job terminates.

You are correct that we have to release partitions for reset vertices, which will be added in a follow-up.